spring-cloud-gcp-pubsub/src/main/java/org/springframework/cloud/gcp/pubsub/support/DefaultPublisherFactory.java

```diff
@@ -56,9 +59,17 @@
 
 	private final CredentialsProvider credentialsProvider;
 
-	public DefaultPublisherFactory(String projectId, ExecutorProvider executorProvider,
-			ChannelProvider channelProvider, CredentialsProvider credentialsProvider) {
-		this.projectId = projectId;
+	public DefaultPublisherFactory(GcpProjectIdProvider projectIdProvider,
+			ExecutorProvider executorProvider,
+			ChannelProvider channelProvider,
+			CredentialsProvider credentialsProvider) {
+		Assert.notNull(projectIdProvider, "The project ID provider can't be null.");
+		Assert.notNull(executorProvider, "The executor provider can't be null.");
+		Assert.notNull(channelProvider, "The channel provider can't be null.");
+		Assert.notNull(credentialsProvider, "The credentials provider can't be null.");
+
+		this.projectId = projectIdProvider.getProjectId()
+				.orElseThrow(MissingProjectIdException::new);
```

Comment:
Ah apologies, I didn't understand your words in that way.

Even though I don't think it's the right thing to not use the cool features that newer versions of Java provide, I'll get rid of the `Optional`.

FWIW, inconsistency within Spring and Spring Boot is not unheard of -- for example, it's my understanding that Spring Kafka provides a lot of best practices that other projects lack. Not innovating just because other projects also don't doesn't sound right to me, but you will eventually own the code, so you have the last say!